openedx/core/djangoapps/oauth_dispatch/docs/decisions/0009-jwt-in-session-cookie.rst

```diff
@@ -0,0 +1,203 @@
+
+#. **Refresh JWT Cookies.** When a JWT expires, we do not want to ask the user to login again while their browser
+   session remains alive. A microfrontend will detect JWT expiration upon receiving a 401 response from an API
+   endpoint, or preemptively recognize an imminent expiration. To automatically refresh the JWT cookie, the
+   microfrontend will call a new endpoint ("refresh") that returns a new JWT Cookie to keep the user's session alive.
+
+   * To support this, the login endpoint will include 3 related cookies in its response:
+
+     * **Two JWT Cookies** (as described above), with a *domain* setting so that it is forwarded to any microservice
+       in the system.
+     * **JWT Refresh Cookie**, with a *domain* setting so that it is sent to the login service only.
+
+#. **Remove JWT Cookie on Logout.** When the user logs out, we will remove all JWT-related cookies in the response,
+   which will remove them from the user's browser cookie jar. Thus, the user will be logged out of all the
+   microfrontends.
+
+.. _`current open edX session cookies that may have no expiration`: https://github.com/edx/edx-platform/blob/92030ea15216a6641c83dd7bb38a9b65112bf31a/common/djangoapps/student/cookies.py#L25-L27
+.. _JWT blacklist: https://auth0.com/blog/blacklist-json-web-token-api-keys/
+.. _`JWT ID (jti)`: http://self-issued.info/docs/draft-ietf-oauth-json-web-token.html#jtiDef
+
+
+JWT Cookie Content
+^^^^^^^^^^^^^^^^^^
+
+#. **Minimize JWT size.** According to the `HTTP Cookie RFC standard`_, HTTP cookies `up to 4096 bytes`_ should be
+   supported by a browser. `Modern browsers have treated this requirement as a maximum`_ - and hence do not support
+   more than 4096 bytes. Our current JWT size is about 970 bytes (varying with size of user identifiers, like user's
+   name, etc). (Side note: Signing a JWT with a 2048 byte asymmetric key increases the JWT's size by 325 bytes.)
+   
+   To minimize the JWT's size from the start, we should eliminate any unnecessary data that is `currently embedded
+   in the JWT`_. For example:
+
+   * *aud* - should remove this since we do not make use of the audience field.
+   * *preferred_username* - should be renamed simply to *username*.
+   * *administrator* - can keep for now, but may eventually be replaced as *role* data - when we design
+     authorization.
+
+.. _HTTP Cookie RFC standard: https://tools.ietf.org/html/rfc6265
+.. _up to 4096 bytes: https://tools.ietf.org/html/rfc6265#section-6.1
+.. _Modern browsers have treated this requirement as a maximum: http://browsercookielimits.squawky.net/
+.. _currently embedded in the JWT: https://github.com/edx/edx-platform/blob/92030ea15216a6641c83dd7bb38a9b65112bf31a/openedx/core/lib/token_utils.py#L13
+
+
+JWT Cookie Security
+^^^^^^^^^^^^^^^^^^^
+
+#. **Enable CSRF Protection.** Storing JWTs in HTTP cookies is potentially vulnerable to CSRF attacks.
```

Comment:
We will need to add something in here which talks about cross domain usage of CSRF tokens across backend services once we settle on how we are going to handle this situation. See https://openedx.atlassian.net/browse/ARCH-239.

Reply:
Update ticket to update the decision record.

Reply:
ARCH-239 was updated.